checker/src/org/checkerframework/checker/upperbound/UpperBoundAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,770 @@
+                return true;
+            } else if (AnnotationUtils.areSameByClass(rhs, LTEqLengthOf.class)) {
+                return false;
+            }
+
+            // Neither is LTEL. Both must be LTOM, LTL, or Bottom. And the two must be
+            // different. The only way this can return true is if rhs is Bottom, or if rhs is
+            // LTOM and lhs is LTL.
+            if (AnnotationUtils.areSameByClass(rhs, UpperBoundBottom.class)) {
+                return true;
+            }
+
+            if (AnnotationUtils.areSameByClass(rhs, LTOMLengthOf.class)
+                    && AnnotationUtils.areSameByClass(lhs, LTLengthOf.class)) {
+                return true;
+            }
+
+            // Every other case results in false.
+            return false;
+        }
+    }
+
+    @Override
+    public TreeAnnotator createTreeAnnotator() {
+        return new ListTreeAnnotator(
+                new UpperBoundTreeAnnotator(this), super.createTreeAnnotator());
+    }
+
+    protected class UpperBoundTreeAnnotator extends TreeAnnotator {
+
+        public UpperBoundTreeAnnotator(UpperBoundAnnotatedTypeFactory factory) {
+            super(factory);
+        }
+
+        /**
+         * This exists specifically for Math.min. We need to special case Math.min because it has
+         * unusual semantics: it can be used to combine annotations for the UBC, so it needs to be
+         * special cased. Do not special case other methods here unless you have a compelling reason
+         * to do so.
+         */
+        @Override
+        public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror type) {
+            ExecutableElement fcnMin =
+                    TreeUtils.getMethod("java.lang.Math", "min", 2, processingEnv);
+            if (TreeUtils.isMethodInvocation(tree, fcnMin, processingEnv)) {
+                AnnotatedTypeMirror leftType = getAnnotatedType(tree.getArguments().get(0));
+                AnnotatedTypeMirror rightType = getAnnotatedType(tree.getArguments().get(1));
+
+                type.replaceAnnotation(
+                        qualHierarchy.greatestLowerBound(
+                                leftType.getAnnotationInHierarchy(UNKNOWN),
+                                rightType.getAnnotationInHierarchy(UNKNOWN)));
+            }
+            return super.visitMethodInvocation(tree, type);
+        }
+
+        @Override
+        public Void visitMemberSelect(MemberSelectTree tree, AnnotatedTypeMirror type) {
+            if (tree.getIdentifier().contentEquals("length")
+                    && InternalUtils.typeOf(tree.getExpression()).getKind() == TypeKind.ARRAY) {
+                String arrName =
+                        FlowExpressions.internalReprOf(this.atypeFactory, tree.getExpression())
+                                .toString();
+                type.replaceAnnotation(
+                        qualHierarchy.greatestLowerBound(
```

Comment:
What if this isn't the first time that this array's length is annotated, though? I think at this point it's likely that I'm misunderstanding something, but I'm thinking of something along these lines:
```java
void test(@LTLengthOf("#2") int x, int[] a, int[] b) {
     if (x == b.length) { 
         // At this point, b.length is @LTEqLengthOf("a", "b") via dataflow refinement
         int y = b.length; // I'm concerned with b.length's type at this point. b.length has 
                                    // already be refined, so if I don't call glb, y will have type 
                                    // @LTEqLengthOf("b") instead of {"a","b"}.


     }
}
```

What am I misunderstanding?